rust/gui-smoke-test/src/main.rs

```diff
@@ -0,0 +1,229 @@
+        .chain(args.iter().copied())
+        .collect();
+        let xvfb = Exec::cmd("xvfb-run").args(&args).to_cmdline_lossy();
+
+        tracing::debug!(?xvfb);
+
+        let cmd = Exec::cmd("sudo") // We need `sudo` to run `su`
+            .args(&[
+                "--preserve-env",
+                "su",      // We need `su` to get a login shell as ourself
+                "--login", // And we need a login shell so that the group membership will take effect immediately
+                "--whitelist-environment=XDG_RUNTIME_DIR",
+                &self.username,
+                "--command",
+                &xvfb,
+            ])
+            .env("WEBKIT_DISABLE_COMPOSITING_MODE", "1"); // Might help with CI
+        Ok(cmd)
+    }
+}
+
+#[cfg(target_os = "windows")]
+impl App {
+    fn new() -> Result<Self> {
+        Ok(Self {})
+    }
+
+    fn crash_dump_path(&self) -> PathBuf {
+        let app_data = std::env::var("LOCALAPPDATA").expect("$LOCALAPPDATA should be set");
+        PathBuf::from(app_data)
+            .join("dev.firezone.client/data/logs")
+            .join("last_crash.dmp")
+    }
+
+    // Strange signature needed to match Linux
+    fn gui_command(&self, args: &[&str]) -> Result<Exec> {
+        Ok(Exec::cmd(gui_path()).args(args))
+    }
+}
+
+fn build_binary(name: &str) -> Result<()> {
+    Exec::cmd("cargo")
+        .args(&["build", "--bin", name])
+        .join()?
+        .fz_exit_ok()?;
+    Ok(())
+}
+
+// Get debug symbols from the exe / pdb
+fn dump_syms() -> Result<()> {
+    Exec::cmd("dump_syms")
+        .args(&[
+            debug_db_path().as_os_str(),
+            gui_path().as_os_str(),
+            OsStr::new("--output"),
+            syms_path().as_os_str(),
+        ])
+        .join()?
+        .fz_exit_ok()?;
+    Ok(())
+}
+
+#[cfg(target_os = "linux")]
+fn debug_db_path() -> PathBuf {
+    Path::new("target").join("debug").join(GUI_NAME)
+}
+
+#[cfg(target_os = "windows")]
+fn debug_db_path() -> PathBuf {
+    Path::new("target")
+        .join("debug")
+        .join("firezone_gui_client.pdb")
+}
+
+#[cfg(target_os = "linux")]
+fn ipc_service_command() -> Exec {
+    Exec::cmd("sudo").args(&[OsStr::new("--preserve-env"), ipc_path().as_os_str()])
+}
+
+#[cfg(target_os = "windows")]
+fn ipc_service_command() -> Exec {
+    Exec::cmd(ipc_path())
+}
+
+// `ExitStatus::exit_ok` is nightly, so we add an equivalent here
+trait ExitStatusExt {
+    fn fz_exit_ok(&self) -> Result<()>;
+}
+
+impl ExitStatusExt for subprocess::ExitStatus {
+    fn fz_exit_ok(&self) -> Result<()> {
+        if !self.success() {
+            bail!("Subprocess should exit with success");
+        }
+        Ok(())
+    }
+}
+
+fn gui_path() -> PathBuf {
+    Path::new("target")
+        .join("debug")
+        .join(GUI_NAME)
+        .with_extension(EXE_EXTENSION)
+}
+
+fn ipc_path() -> PathBuf {
+    Path::new("target")
+        .join("debug")
+        .join(IPC_NAME)
+        .with_extension(EXE_EXTENSION)
+}
```

Comment:
I'm probably missing something, but can these just be constants too since they seem to just return static data?

Reply:
`.join` uses the correct platform-separator and is a non-const function which I guess is the reason they aren't const.